README.md

```diff
@@ -1,111 +1,170 @@
 
+### Setup
+
 1. Clone this repository to your local machine.
-2. Copy the `env.example` to a `.env` file on the root directory
-3. Install dependencies by running:
+2. Copy the `env.example` file to `.env` in the root directory.
+3. Install dependencies:
 
-```bash
-npm install
-```
+    ```bash
+    npm install
+    ```
 
-### Database setup
+### Database Setup
 
-1. Run local postgres instance (currently no initial data is inserted)
+1. Run a local PostgreSQL instance (no initial data is inserted):
 
-```shell
-docker-compose up db -d
-```
+    ```shell
+    docker-compose up db -d
+    ```
 
 2. Deploy the contracts locally:
 
-```bash
-npm run contracts start:db
-```
+    ```bash
+    npm run contracts start:db
+    ```
+
+3. Start the gateway:
+
+    ```bash
+    npm run gateway dev:db
+    ```
 
-3. Start the gateway
+4. Write properties to a given domain:
 
-```bash
-npm run gateway dev:db
-```
+    ```bash
+      npm run client start:write:db
+    ```
 
-4. Request the domain properties through the client:
+5. Request domain properties through the client:
 
-```bash
-npm run client start
-```
+    ```bash
+    npm run client read
+    ```
 
-### Layer 2 setup
+### Layer 2 Setup
 
-1. Deploy the contracts locally:
+1. Deploy the contracts to the local Arbitrum node (follow the [Arbitrum's local node setup tutorial](https://docs.arbitrum.io/run-arbitrum-node/run-local-dev-node)):
 
-```bash
-npm run contracts start:arb
-```
+    ```bash
+    npm run contracts start:arb:l2
+    ```
 
-2. Start the gateway
+2. Gather the contract address from the terminal and add it [here](https://github.com/blockful-io/external-resolver/blob/main/packages/contracts/script/local/ArbResolver.s.sol#L56) so the L1 domain gets resolved by the L2 contract you just deployed.
 
-```bash
-npm run gateway dev:arb
-```
+3. Start the gateway:
 
-3. Request the domain properties through the client:
+    ```bash
+    npm run gateway dev:arb
+    ```
 
-```bash
-npm run client start
-```
+4. Request domain properties through the client:
+
+    ```bash
+    npm run client start
+    ```
 
 ## Deployment
 
-It's required to have the [Railway CLI](https://docs.railway.app/guides/cli) installed.
+### Prerequisites
+
+Ensure you have the [Railway CLI](https://docs.railway.app/guides/cli) installed.
+
+1. Install the Railway CLI:
+
+    ```bash
+    npm i -g @railway/cli
+    ```
 
-1. Install the Railway CLI through: `npm i -g @railway/cli`
-2. Login into your Railway account: `railway login`
-3. Link the repo to the project:  `railway link`
-4. Build project: `turbo build --filter=@blockful/gateway...`
-5. Test bundle locally: `npm run gateway start:db`
-6. Deploy the Gateway: `railway up`
+2. Log in to your Railway account:
+
+    ```bash
+    railway login
+    ```
+
+3. Link the repo to the project:
+
+    ```bash
+    railway link
+    ```
+
+4. Deploy the Gateway:
+
+    ```bash
+    railway up
+    ```
 
 ## Architecture
```

Comment:
On this architecture section, it would be good if we have an explanation for each diagram. They are self-explanatory for us that already know ENS's and the Offchain Resolution's structure. But, maybe, it is not that clear for some one that is falling by parachute and wants to know what we are doing.